test/cpp/api/serialize.cpp

```diff
@@ -33,6 +35,137 @@ torch::Tensor save_and_load(torch::Tensor input) {
 }
 } // namespace
 
+template <typename DerivedOptions>
+bool is_optimizer_param_groups_equal(const OptimizerParamGroup& lhs, const OptimizerParamGroup& rhs) {
+  const auto& lhs_params= lhs.params();
+  const auto& rhs_params = rhs.params();
+  if(lhs_params.size() != rhs_params.size()) {
+    return false;
+  }
+  for (int64_t j=0; j<lhs_params.size(); j++) {
+    if(!torch::equal(lhs_params[j], rhs_params[j]))
+      return false;
+  }
+  if(!(static_cast<const DerivedOptions&>(lhs.options()) == static_cast<const DerivedOptions&>(rhs.options()))) {
+    return false;
+  }
+  return true;
+}
+
+template <typename DerivedOptimizerParamState>
+bool is_optimizer_state_equal(
+  const std::vector<torch::Tensor>& lhs_params,
+  const ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& lhs_state,
+  const std::vector<torch::Tensor>& rhs_params,
+  const ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& rhs_state) {
```

Comment:
Some thought: maybe we can change this function to check whether `lhs_state` and `rhs_state` are exactly the same? In that case, we probably don't need to take `lhs_params` and `rhs_params` as input arguments, and in the implementation we can take the keys from e.g. `lhs_state` and check that the corresponding values are equal from both states.

Reply:
Umm I don't think that would help to compare two hash maps that ways because the keys in hash map can be stored in any order

Reply:
do you mean using == overload fn for hash_maps? will (value.second != found->second) work if we define a != overload for states?
`auto found = rhs.find(value.first);
            if (found == rhs.end())
                return false;
            else if (value.second != found->second)
                return false;`

Reply:
yes I agree will do :D thanks!